### PR TITLE
Cycle prime decomposition is closed under cartesian product

### PR DESCRIPTION
**Pull request by VictorBlanchi**
In this pull request, I wanted to prove that the cartesian product of the cycles prime decomposition of `n` and `m` is equal to the cycle prime decomposition of `mul-\bN n m`. Initially, I thought it would be easy by using the fundamental theorem of arithmetic but I encountered some unexpected problems that I had to solve ^^. 
The main idea was to 

1. Show that if we take the sorted concatenation of the lists corresponding to the prime decomposition of `n`  and `m` we obtain a prime decomposition of `mul-\bN n m`. This was harder than I thought. I had to show that `mul-list-\bN` is invariant by permuting the list. I made the general case, and to do it I had to show that every transposition (on Fin n) is a composition of adjacent transpositions.
2. Then I used the fundamental theorem of arithmetic to demonstrate that the sorted concatenation of the lists corresponding to the prime decomposition of `n` and `m` is equal to the prime decomposition of `mul-\bN n m`.
3. Finally, I expected to only need to permute the cycles in the cartesian product of the cycles prime decomposition associated to `n` and `m` with the permutation given by the sort by insertion. I thought that applying step 2. would be sufficient to conclude. But I had to show this property : `map-list f (permute-list p t)` is equal to `permute-list (map-list p f) t` and which was quite hard to prove.

### CODE REVIEW

File: src/elementary-number-theory/fundamental-theorem-of-arithmetic.lagda.md [hunk 1]
```diff
@@ -934,3 +940,138 @@ pr2 (fundamental-theorem-arithmetic-list-ℕ x H) d =
        ( is-prime-decomposition-list-fundamental-theorem-arithmetic-ℕ x H)
        ( pr2 d))
 ```
+
+### The sorted list associated with the concatenation of the prime decomposition of `n` and the prime decomposition of `m` is the prime decomposition of `mul-ℕ n m`
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/fundamental-theorem-of-arithmetic.lagda.md:944]: Just a heads up:
We introduced the infix binary operator `_*ℕ_` for the multiplication of natural numbers (and similar for other standard types) in a very recent PR (#620). Feel free to start using this notation if you want. Although it is not required, it is preferred in most cases now I think (a bit like `Id` vs. `=`).

Comment (fredrik-bakke) [reply src/elementary-number-theory/fundamental-theorem-of-arithmetic.lagda.md:944]: Sorry, I should have written this as a general comment on the PR. You can go ahead and resolve this comment regardless of your choice. 😅

File: src/lists/permutation-lists.lagda.md [hunk 2]
```diff
@@ -36,7 +49,265 @@ module _
...
+      ( inv (length-permute-list p t))
+      ( ( helper-compute-list-vec-map-vec-permute-vec-vec-list f p t)))
+
+private
+  -- I think that this lemma is already in the library, but I didn't find it
+  lemma-tr :
+    {l1 l2 : Level} {A : UU l1} {B : A → UU l2}
+    (x y : A) (p : x ＝ y) (b : B y) →
+    tr B p (tr B (inv p) b) ＝ b
+  lemma-tr x y refl b = refl
```
Comment (fredrik-bakke) [inline src/lists/permutation-lists.lagda.md:275]: This is `issec-inv-tr` from `foundation.identity-types`. It took me a bit of time to find it as well 😅

File: src/lists/permutation-vectors.lagda.md [hunk 3]
```diff
@@ -107,3 +153,244 @@ module _
...
+          ( isretr-functional-vec-vec n (functional-vec-vec n v ∘ map-equiv t)))
+        ( is-in-permute-functional-vec-is-in-functional-vec
+          ( n)
+          ( functional-vec-vec n v)
+          ( t)
+          ( x)
+          ( is-in-functional-vec-is-in-vec n v x I)))
+```
+
+### If `(μ : A → (B → B))` satisfies a property of commutativity, then for `b : B` the function `fold-vec b μ` is invariant by permutation
```
Comment (fredrik-bakke) [inline src/lists/permutation-vectors.lagda.md:210]:
```
### If `(μ : A → (B → B))` satisfies a commutativity property, `fold-vec b μ` is invariant under permutation for every `b : B`
```

File: src/lists/sorting-algorithms-lists.lagda.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+open import order-theory.decidable-total-orders
+
+open import univalent-combinatorics.standard-finite-types
+```
+
+</details>
+
+## Idea
+
+In these file we define the notion of sorting algorithms.
```
Comment (fredrik-bakke) [inline src/lists/sorting-algorithms-lists.lagda.md:36]: In <ins>this</ins> file we define the notion of sorting algorithms <ins>for lists</ins>.

File: src/univalent-combinatorics/2-element-decidable-subtypes.lagda.md [hunk 5]
```diff
@@ -26,6 +26,7 @@ open import foundation.functions
 open import foundation.functoriality-coproduct-types
 open import foundation.functoriality-dependent-pair-types
 open import foundation.functoriality-propositional-truncation
+open import foundation.homotopies
```
Comment (fredrik-bakke) [inline src/univalent-combinatorics/2-element-decidable-subtypes.lagda.md:29]: It looks like this import is not needed.

File: src/univalent-combinatorics/cycle-prime-decomposition-natural-numbers.lagda.md [hunk 6]
```diff
@@ -55,3 +69,97 @@ cycle-prime-decomposition-ℕ n H =
   iterated-product-lists
     ( map-list (Cyclic-Type lzero) (list-fundamental-theorem-arithmetic-ℕ n H))
 ```
+
+## Properties
+
+### Cycle prime decomposition is closed by cartesian product
```
Comment (fredrik-bakke) [inline src/univalent-combinatorics/cycle-prime-decomposition-natural-numbers.lagda.md:75]: Cycle prime decomposition<ins>s</ins> <ins>are</ins> closed <ins>under</ins> cartesian product<ins>s</ins>

File: src/univalent-combinatorics/cycle-prime-decomposition-natural-numbers.lagda.md [hunk 7]
```diff
@@ -55,3 +69,97 @@ cycle-prime-decomposition-ℕ n H =
...
+
+The cartesian product of the cycle prime decomposition of `n` and `m` is equal
+to the cycle prime decomposition of `mul-ℕ n m`
```
Comment (fredrik-bakke) [inline src/univalent-combinatorics/cycle-prime-decomposition-natural-numbers.lagda.md:78]: Add punctuation at the end.

**Review by fredrik-bakke — APPROVED** [review]
Overall an excellent PR! 😊

**Comment by fredrik-bakke** [comment]
Excellent! Looks ready to be merged to me now, but again I'll wait and see if Egbert has anything to add.

**Comment by EgbertRijke** [comment]
Yep, let's merge

**Pull request merged**